solver/bboltcachestorage/storage.go

```diff
@@ -455,3 +451,55 @@ func isEmptyBucket(b *bolt.Bucket) bool {
 	k, _ := b.Cursor().First()
 	return k == nil
 }
+
+// safeOpenDB opens a bolt database and recovers from panic that
+// can be caused by a corrupted database file.
+func safeOpenDB(dbPath string) (db *bolt.DB, err error) {
+	defer fallbackOpenDB(dbPath, &db, &err)
+	return openDB(dbPath)
+}
+
+// fallbackOpenDB performs the panic recovery, database backup, and
+// opening of the new database file when a panic happens from safeOpenDB.
+func fallbackOpenDB(dbPath string, db **bolt.DB, err *error) {
+	r := recover()
+	if r == nil {
+		return
+	}
+	// Wrap the recover error with errorf so we get the stack trace in the error.
+	recoverErr := errors.Errorf("%v", r)
+
+	backupPath := dbPath + "." + identity.NewID() + ".bak"
+	bklog.L.Errorf("failed to open database file %s, resetting to empty. Old database is backed up to %s. "+
+		"This error signifies that buildkitd likely crashed or was sigkilled abrubtly, leaving the database corrupted. "+
+		"If you see logs from a previous panic then please report in the issue tracker at https://github.com/moby/buildkit . %+v", dbPath, backupPath, recoverErr)
+	if renameErr := os.Rename(dbPath, backupPath); renameErr != nil {
+		*err = errors.Wrapf(renameErr, "failed to rename database file %s to %s", dbPath, backupPath)
+		return
+	}
+
+	// Attempt to open the database again. This should be a new database.
+	*db, *err = openDB(dbPath)
+	return
+}
+
+// openDB opens a bolt database in user-only read/write mode.
+func openDB(dbPath string) (*bolt.DB, error) {
+	return bolt.Open(dbPath, 0600, nil)
+}
+
+// initStore initializes the store with the required buckets.
+func initStore(db *bolt.DB) error {
```

Comment:
Just thought it would make things easier to read while I was modifying things. I can revert it if you prefer but I thought it would be easier to read.

Reply:
It is a long function (yet) and only has one caller so I don't think there is a need to slice it. There is also no scope difference as "new" and "init" mean the same thing, just with a different word.

Reply:
I moved this back. It's not the primary purpose of this PR and the initialization isn't really that complicated anyway.